nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -148,57 +156,81 @@ void KeepoutFilter::process(
     return;
   }
 
-  double wx, wy;  // world coordinates
-
-  // Optimization: iterate only in overlapped
-  // (min_i, min_j)..(max_i, max_j) & mask_costmap_ area.
-  //
-  //           mask_costmap_
-  //       *----------------------------*
-  //       |                            |
-  //       |                            |
-  //       |      (2)                   |
-  // *-----+-------*                    |
-  // |     |///////|<- overlapped area  |
-  // |     |///////|   to iterate in    |
-  // |     *-------+--------------------*
-  // |    (1)      |
-  // |             |
-  // *-------------*
-  //  master_grid (min_i, min_j)..(max_i, max_j) window
-  //
-  // ToDo: after costmap rotation will be added, this should be re-worked.
-
-  // Calculating bounds corresponding to bottom-left overlapping (1) corner
-  int mg_min_x, mg_min_y;  // masger_grid indexes of bottom-left (1) corner
-  // mask_costmap_ -> master_grid intexes conversion
-  const double half_cell_size = 0.5 * mask_costmap_->getResolution();
-  wx = mask_costmap_->getOriginX() + half_cell_size;
-  wy = mask_costmap_->getOriginY() + half_cell_size;
-  master_grid.worldToMapNoBounds(wx, wy, mg_min_x, mg_min_y);
-  // Calculation of (1) corner bounds
-  if (mg_min_x >= max_i || mg_min_y >= max_j) {
-    // There is no overlapping. Do nothing.
-    return;
-  }
-  mg_min_x = std::max(min_i, mg_min_x);
-  mg_min_y = std::max(min_j, mg_min_y);
-
-  // Calculating bounds corresponding to top-right window (2) corner
-  int mg_max_x, mg_max_y;  // masger_grid indexes of top-right (2) corner
-  // mask_costmap_ -> master_grid intexes conversion
-  wx = mask_costmap_->getOriginX() +
-    mask_costmap_->getSizeInCellsX() * mask_costmap_->getResolution() + half_cell_size;
-  wy = mask_costmap_->getOriginY() +
-    mask_costmap_->getSizeInCellsY() * mask_costmap_->getResolution() + half_cell_size;
-  master_grid.worldToMapNoBounds(wx, wy, mg_max_x, mg_max_y);
-  // Calculation of (2) corner bounds
-  if (mg_max_x <= min_i || mg_max_y <= min_j) {
-    // There is no overlapping. Do nothing.
-    return;
+  tf2::Transform tf2_transform;
+  tf2_transform.setIdentity();  // initialize by identical transform
+  int mg_min_x, mg_min_y;  // masger_grid indexes of bottom-left window corner
+  int mg_max_x, mg_max_y;  // masger_grid indexes of top-right window corner
+
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // prepare frame transformation if mask_frame_ != global_frame_
+    geometry_msgs::msg::TransformStamped transform;
+    try {
+      transform = tf_->lookupTransform(
+        mask_frame_, global_frame_, tf2::TimePointZero,
+        transform_tolerance_);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR(node->get_logger(), "KeepoutFilter: %s", ex.what());
```

Comment:
A contextual error would be nice, adding something like "KeepoutFilter: failed to get mask frame %s transformation to %s frame with error: %s" with the frame names listed.

Reply:
Done in https://github.com/ros-planning/navigation2/pull/2054/commits/f5b19ebd9d0517f46339632d5058f3d2202b4fd0